### PR TITLE
Macros

### PR DESCRIPTION
**Pull request by alexcarsello**
This PR adds support for macros using the `--macros` flag.

Given a path (`--macros some/macro/dir`) PDQ will recursively search for all macro files (*.db, *.lef, *.v) at that path, add them to the mflowgen flow, and connect them to the nodes that require them.

### CODE REVIEW

File: basic_flow/construct.py.tpl [hunk 1]
```diff
@@ -62,11 +63,23 @@ def construct():
   sim         = Step( 'synopsys-vcs-sim',               default=True )
   ptpx_gl     = Step( 'synopsys-ptpx-gl',    default=True )
 
+  # Add inputs/outputs for all macros
+  macro_files = {{ macro_files }}
+  macro_db = [f for f in macro_files if '.db' in f]
+  macro_lef = [f for f in macro_files if '.lef' in f]
+  macro_v = [f for f in macro_files if '.v' in f]
```
Comment (rsetaluri) [inline basic_flow/construct.py.tpl]: lets use `str.endswith` (https://python-reference.readthedocs.io/en/latest/docs/str/endswith.html)

File: basic_flow/macros/configure.yml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+name: macros
+
+#=========================================================================
+# Commands
+#=========================================================================
+
+commands:
+  - mkdir -p outputs && cd outputs
+  - ln -sf ../*.db .
```
Comment (rsetaluri) [inline basic_flow/macros/configure.yml:13]: symbolic link seems reasonable

File: pdq/flow_tools/basic_flow.py [hunk 3]
```diff
@@ -18,9 +19,35 @@ class BasicFlowOpts:
...
+        if path.lower().endswith(('.db', '.lef', '.v')):
+            #FileCopy(path, builder.get_relative("macros/outputs/"))
+            macro_file_list.append(path)
+    # If it's a directory, recursively search for macro files
+    elif os.path.isdir(path):
+        for f in os.listdir(path):
+            macro_file_list += get_macro_files(os.path.join(path, f))
+
+    return macro_file_list
+
```
Comment (rsetaluri) [inline pdq/flow_tools/basic_flow.py]:
```suggestion
def _get_macro_files(path):
    """Takes in a path and gets all macro files present there."""
    macro_file_list = []
    # If it's a file, copy it and add it to the filename list.
    if os.path.isfile(path):
        if path.lower().endswith(('.db', '.lef', '.v')):
            macro_file_list.append(path)
    # If it's a directory, recursively search for macro files.
    elif os.path.isdir(path):
        for f in os.listdir(path):
            macro_file_list += get_macro_files(os.path.join(path, f))

    return macro_file_list
```

File: pdq/flow_tools/basic_flow.py [hunk 4]
```diff
@@ -18,9 +19,35 @@ class BasicFlowOpts:
...
 
 
 def make_basic_flow(ckt: m.DefineCircuitKind, opts: BasicFlowOpts):
+    # First, parse the macro args and copy all the files to the macro node
+    macro_path_list = [s.strip() for s in opts.macros.split(",")]
+    macro_file_list = []
+    for path in macro_path_list:
+        macro_file_list += get_macro_files(path)
+
+    macro_file_basenames = [os.path.basename(f) for f in macro_file_list]
```
Comment (rsetaluri) [inline pdq/flow_tools/basic_flow.py]:
```suggestion
    # First, parse the macro args and copy all the files to the macro node.
    macro_path_list = [s.strip() for s in opts.macros.split(",")]
    macro_file_list = []
    for path in macro_path_list:
        macro_file_list += _get_macro_files(path)

    macro_file_basenames = [os.path.basename(f) for f in macro_file_list]
```

**Review by rsetaluri — APPROVED** [review]
few minor comments. looks good otherwise

File: pdq/flow_tools/basic_flow.py (hunk 3 above)
Comment (alexcarsello) [reply pdq/flow_tools/basic_flow.py]: Ah I see I left an old commented out line there. But what's the reason for putting a comment in triple quotes?

Comment (alexcarsello) [reply pdq/flow_tools/basic_flow.py]: Oh it's a docstring... got it.

File: basic_flow/construct.py.tpl (hunk 1 above)
Comment (alexcarsello) [reply basic_flow/construct.py.tpl]: done

**Pull request merged**